bin/dev-tools/create_shipments.py

```diff
@@ -0,0 +1,332 @@
+            self.active_users.add(user)
+        return self.users[user]
+
+    def _validate_attributes(self, attributes):
+        if self.verbose:
+            print(f'Validating attributes: {attributes}')
+        for attribute in attributes:
+            try:
+                attribute = json.loads(attribute)
+            except JSONDecodeError:
+                if self.verbose:
+                    print(f'Non json attribute recieved: {attribute}')
+            if isinstance(attribute, dict):
+                self.attributes.update(attribute)
+            else:
+                try:
+                    key, value = attribute.split("=")
+                except ValueError:
+                    self._process_failure(f'Invalid format for attribute: {attribute}, should be in format: key=value')
+                self.attributes[key] = value
+
+    def _process_failure(self, error_message):
+        print(f'Shipment Creator failed: {error_message}')
+        print(f'Response errors: {self.errors}')
+        sys.exit(2)
+
+    def _parse_request(self, url, method='post', **kwargs):
+        if method not in ['post', 'get', 'patch']:
+            self._process_failure(f'Invalid method: {method}')
+        request_method = getattr(requests, method)
+        try:
+            response = request_method(url, **kwargs)
+        except requests.exceptions.ConnectionError:
+            if self.verbose:
+                print(f'Connection error raised when connecting to {url}')
+            self._process_failure(f'Error connecting to url: {url}')
+
+        if not response.ok:
+            if self.verbose:
+                print(f'Invalid response returned from {url}')
+            self.errors += response.json()['errors']
+            return None
+        return response.json()
+
+    def _retrieve_updated_attributes(self):
+        if self.verbose:
+            print(f'Updating attributes to remove update random and sequential variables')
+        updated_attributes = deepcopy(self.attributes)
+        for key, value in updated_attributes.items():
+            if value == "##RAND##":
+                updated_attributes[key] = str(uuid4())
+            elif value == "##NUM##":
+                updated_attributes[key] = self.sequence_number
+        if self.device:
+            device_response = self._parse_request(
+                f"{self.profiles_url}/api/v1/device",
+                data={'device_type': "AXLE_GATEWAY"},
+                headers={'Authorization': 'JWT {}'.format(self.shipper['token'])})
+            if not device_response:
+                self._process_failure(f'Error generating device')
+            updated_attributes['device_id'] = device_response['data']['id']
+        self.sequence_number += 1
+        return updated_attributes
+
+    def _set_user_tokens(self):
+        for user in self.active_users:
+            if self.users[user]['token_exp'] and (self.users[user]['token_exp'] < datetime.now(timezone.utc)):
```

Comment:
In the case where the token expires in the _very near_ future, it is possible this will not refresh a token that will be expired by the time we use it.  Including a buffer in the calculation here will save you on edge case headaches in the future.

Reply:
Is 15 seconds enough buffer time you think?

Reply:
Token TTL is 5 minutes.  If I were doing this, I'd probably refresh when <60 seconds remained.